pyro/contrib/conjugate/infer.py

```diff
@@ -0,0 +1,164 @@
+
+
+def _compound(base, parent):
+    return _make_cls(base, {}, {"parent": parent})
+
+
+class BetaBinomialPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Beta, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Binomial, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration1 = self._latent.concentration1
+        concentration0 = self._latent.concentration0
+        total_count = self._conditional.total_count
+        reduce_dims = len(obs.size()) - len(concentration1.size())
+        # Unexpand total_count to have the same shape as concentration0.
+        # Raise exception if this isn't possible.
+        total_count = sum_leftmost(total_count, reduce_dims)
+        summed_obs = sum_leftmost(obs, reduce_dims)
+        return dist.Beta(concentration1 + summed_obs,
+                         total_count + concentration0 - summed_obs,
+                         validate_args=self._latent._validate_args)
+
+    def compound(self):
+        return _compound(dist.BetaBinomial, parent=self)(concentration1=self._latent.concentration1,
+                                                         concentration0=self._latent.concentration0,
+                                                         total_count=self._conditional.total_count)
+
+
+class GammaPoissonPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Gamma, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Poisson, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration = self._latent.concentration
+        rate = self._latent.rate
+        reduce_dims = len(obs.size()) - len(rate.size())
+        num_obs = obs.shape[:reduce_dims].numel()
+        summed_obs = sum_leftmost(obs, reduce_dims)
+        return dist.Gamma(concentration + summed_obs, rate + num_obs)
+
+    def compound(self):
+        return _compound(dist.GammaPoisson, parent=self)(concentration=self._latent.concentration,
+                                                         rate=self._latent.rate)
+
+
+class UncollapseConjugateMessenger(ReplayMessenger):
+    r"""
+    Extends `~pyro.poutine.replay_messenger.ReplayMessenger` to uncollapse
+    compound distributions.
+    """
+    def _pyro_sample(self, msg):
+        is_collapsible = getattr(msg["fn"], "collapsible", False)
+        if is_collapsible:
+            observed_node, conj_pair = None, None
```

Comment:
Hmm wouldn't this still work if the conditional site were sampled rather than observed but if its `msg["value"]` had generated a sample that was later used in a downstream likelihood? This may be rare, but I see no reason to hard-code to observed sites here.

Reply:
I suppose I was tunnel-visioned into making this work for HMC, where both the Poisson and Binomial need to be observed as they are discrete variables. I think this should work other-wise too. Let me try to refactor, and update.